File: app/src/main/java/io/github/ovso/psytest/ui/base/interfaces/OnEndlessRecyclerScrollListener.java
package io.github.ovso.psytest.ui.base.interfaces;

import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import lombok.Setter;
import lombok.experimental.Accessors;

public class OnEndlessRecyclerScrollListener extends RecyclerView.OnScrollListener {

  private int visibleThreshold = 1;
  private int lastVisibleItem, totalItemCount;
  private boolean loading;
  private LinearLayoutManager linearLayoutManager;
  private OnLoadMoreListener onLoadMoreListener;

  OnEndlessRecyclerScrollListener(OnEndlessRecyclerScrollListener.Builder builder) {
    linearLayoutManager = builder.linearLayoutManager;
    onLoadMoreListener = builder.onLoadMoreListener;
  }

  @Override public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
    super.onScrolled(recyclerView, dx, dy);

    totalItemCount = linearLayoutManager.getItemCount();
    lastVisibleItem = linearLayoutManager.findLastVisibleItemPosition();
    if (!loading && totalItemCount <= (lastVisibleItem + visibleThreshold)) {
      loading = true;
      if (onLoadMoreListener != null) {
        onLoadMoreListener.onLoadMore();
      }
    }
  }

  public void setLoaded() {
    loading = false;
  }

  public interface OnLoadMoreListener {
    void onLoadMore();
  }

  public static class Builder
      implements io.github.ovso.psytest.ui.base.Builder<OnEndlessRecyclerScrollListener> {
    @Setter @Accessors(chain = true) private LinearLayoutManager linearLayoutManager;
    @Setter @Accessors(chain = true) private OnLoadMoreListener onLoadMoreListener;

    @Override public OnEndlessRecyclerScrollListener build() {
      return new OnEndlessRecyclerScrollListener(this);
    }
  }
}

File: app/src/main/java/io/github/ovso/psytest/ui/base/interfaces/OnEndlessRecyclerScrollListener.kt
package io.github.ovso.psytest.ui.base.interfaces

import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import io.github.ovso.psytest.ui.base.IBuilder

class OnEndlessRecyclerScrollListener private constructor(builder: OnEndlessRecyclerScrollListener.Builder) :
    RecyclerView.OnScrollListener() {

  private var visibleThreshold = 1
  private var lastVisibleItem: Int = 0
  private var totalItemCount: Int = 0
  private var loading: Boolean = false
  private val linearLayoutManager: LinearLayoutManager?
  private val onLoadMoreListener: OnLoadMoreListener?

  init {
    linearLayoutManager = builder.linearLayoutManager
    onLoadMoreListener = builder.onLoadMoreListener
    visibleThreshold = builder.visibleThreshold
  }

  override fun onScrolled(
    recyclerView: RecyclerView,
    dx: Int,
    dy: Int
  ) {
    super.onScrolled(recyclerView, dx, dy)

    totalItemCount = linearLayoutManager!!.itemCount
    lastVisibleItem = linearLayoutManager.findLastVisibleItemPosition()
    if (!loading && totalItemCount <= lastVisibleItem + visibleThreshold) {
      loading = true
      onLoadMoreListener?.onLoadMore()
    }
  }

  fun setLoaded() {
    loading = false
  }

  interface OnLoadMoreListener {
    fun onLoadMore()
  }

  class Builder : IBuilder<OnEndlessRecyclerScrollListener> {
    var visibleThreshold = 1
    var linearLayoutManager: LinearLayoutManager? = null
    var onLoadMoreListener: OnLoadMoreListener? = null
    fun setLayoutManager(layoutManager: LinearLayoutManager): Builder {
      this.linearLayoutManager = layoutManager
      return this
    }

    fun setVisibleThreshold(visibleThreshold: Int): Builder {
      this.visibleThreshold = visibleThreshold
      return this
    }

    fun setOnLoadMoreListener(onLoadMoreListener: OnLoadMoreListener): Builder {
      this.onLoadMoreListener = onLoadMoreListener
      return this
    }

    override fun build(): OnEndlessRecyclerScrollListener {
      return OnEndlessRecyclerScrollListener(this)
    }
  }
}